File: examples/python-deposit-redeem-example/deposit.py
import os
import sys
import json
from dotenv import load_dotenv
from web3 import Web3

load_dotenv()


def get_env():
    rwa_token = os.getenv("RWA_TOKEN_ADDRESS")
    whitelist = os.getenv("WHITELIST_ADDRESS")
    fee_address = os.getenv("FEE_ADDRESS") or ""

    env = {
        "RPC_URL": os.getenv("RPC_URL"),
        "PRIVATE_KEY": os.getenv("PRIVATE_KEY"),
        "RWA_TOKEN_ADDRESS": Web3.to_checksum_address(rwa_token) if rwa_token else "",
        "WHITELIST_ADDRESS": Web3.to_checksum_address(whitelist) if whitelist else "",
        "FEE_ADDRESS": Web3.to_checksum_address(fee_address) if fee_address else "",
        "APPROVE_MAX": os.getenv("APPROVE_MAX", "false").lower() == "true",
        "ENTRY_FEE_TIMING": int(os.getenv("ENTRY_FEE_TIMING") or 0),
    }

    if not all(
        [
            env["RPC_URL"],
            env["PRIVATE_KEY"],
            env["RWA_TOKEN_ADDRESS"],
            env["WHITELIST_ADDRESS"],
        ]
    ):
        raise ValueError("Missing required env values. Please check .env.example")

    return env


def load_abi(filename):
    path = os.path.join(os.path.dirname(__file__), "abis", filename)
    with open(path, "r") as f:
        data = json.load(f)
        return data.get("abi", data)


def resolve_fee_address(vault, env):
    if env["FEE_ADDRESS"]:
        return env["FEE_ADDRESS"]

    fee_address = vault.functions.rwaFee().call()
    if not fee_address or fee_address == "0x0000000000000000000000000000000000000000":
        raise ValueError(
            "Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address."
        )

    return fee_address


def main():
    if len(sys.argv) < 3:
        raise ValueError("Usage: python deposit.py <desiredAssetsBeforeFee>")

    desired_assets_input = sys.argv[2]
    env = get_env()

    w3 = Web3(Web3.HTTPProvider(env["RPC_URL"]))
    account = w3.eth.account.from_key(env["PRIVATE_KEY"])
    user = account.address

    rwa_abi = load_abi("RWAToken.json")
    whitelist_abi = load_abi("Whitelist.json")
    fee_abi = load_abi("RWAFee.json")

    vault = w3.eth.contract(address=env["RWA_TOKEN_ADDRESS"], abi=rwa_abi)
    whitelist = w3.eth.contract(address=env["WHITELIST_ADDRESS"], abi=whitelist_abi)

    print("=== DEPOSIT FLOW (WITH ENTRY FEE) ===")
    print("User:", user)
    print("Vault:", env["RWA_TOKEN_ADDRESS"])
    print("Whitelist:", env["WHITELIST_ADDRESS"])

    is_whitelisted = whitelist.functions.balanceOf(user, 1).call()
    print("Whitelisted:", is_whitelisted)
    if not is_whitelisted:
        raise ValueError("User is not whitelisted.")

    asset_address = vault.functions.asset().call()

    asset = w3.eth.contract(
        address=asset_address,
        abi=[
            {
                "inputs": [],
                "name": "decimals",
                "outputs": [{"type": "uint8"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [],
                "name": "symbol",
                "outputs": [{"type": "string"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [{"type": "address"}],
                "name": "balanceOf",
                "outputs": [{"type": "uint256"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [{"type": "address"}, {"type": "address"}],
                "name": "allowance",
                "outputs": [{"type": "uint256"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [{"type": "address"}, {"type": "uint256"}],
                "name": "approve",
                "outputs": [{"type": "bool"}],
                "stateMutability": "nonpayable",
                "type": "function",
            },
        ],
    )

    asset_decimals = asset.functions.decimals().call()
    asset_symbol = asset.functions.symbol().call()
    desired_assets = int(float(desired_assets_input) * (10**asset_decimals))

    fee_address = resolve_fee_address(vault, env)
    fee_contract = w3.eth.contract(address=fee_address, abi=fee_abi)

    entry_fee = fee_contract.functions.feeOnRaw(
        desired_assets, env["ENTRY_FEE_TIMING"]
    ).call()
    total_spend = desired_assets + entry_fee
    preview_shares = vault.functions.previewDeposit(total_spend).call()

    share_decimals = vault.functions.decimals().call()
    share_symbol = vault.functions.symbol().call()

    print("Fee contract:", fee_address)
    print("Desired assets before fee:", desired_assets_input, asset_symbol)
    print("Entry fee:", entry_fee / (10**asset_decimals), asset_symbol)
    print(
        "Total spend sent to deposit():",
        total_spend / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "previewDeposit(totalSpend):",
        preview_shares / (10**share_decimals),
        share_symbol,
    )

    asset_balance = asset.functions.balanceOf(user).call()
    allowance = asset.functions.allowance(user, env["RWA_TOKEN_ADDRESS"]).call()

    print("Wallet asset balance:", asset_balance / (10**asset_decimals), asset_symbol)
    print("Allowance to vault:", allowance / (10**asset_decimals), asset_symbol)

    if asset_balance < total_spend:
        raise ValueError("Not enough asset balance for desiredAssets + entryFee.")

    if allowance < total_spend:
        approve_amount = 2**256 - 1 if env["APPROVE_MAX"] else total_spend
        print("Allowance is not enough, sending approve()...")
        nonce = w3.eth.get_transaction_count(user)
        approve_tx = asset.functions.approve(
            env["RWA_TOKEN_ADDRESS"], approve_amount
        ).build_transaction(
            {"from": user, "nonce": nonce, "gas": 100000, "gasPrice": w3.eth.gas_price}
        )
        signed_approve_tx = w3.eth.account.sign_transaction(
            approve_tx, env["PRIVATE_KEY"]
        )
        approve_tx_hash = w3.eth.send_raw_transaction(signed_approve_tx.raw_transaction)
        print("Approve tx:", w3.to_hex(approve_tx_hash))
        w3.eth.wait_for_transaction_receipt(approve_tx_hash)
        print("Approve confirmed.")
    else:
        print("Allowance is already enough.")

    print("Sending deposit(totalSpend, receiver)...")
    nonce = w3.eth.get_transaction_count(user)
    deposit_tx = vault.functions.deposit(total_spend, user).build_transaction(
        {"from": user, "nonce": nonce, "gas": 200000, "gasPrice": w3.eth.gas_price}
    )
    signed_deposit_tx = w3.eth.account.sign_transaction(deposit_tx, env["PRIVATE_KEY"])
    deposit_tx_hash = w3.eth.send_raw_transaction(signed_deposit_tx.raw_transaction)
    print("Deposit tx:", w3.to_hex(deposit_tx_hash))
    receipt = w3.eth.wait_for_transaction_receipt(deposit_tx_hash)
    print("Deposit confirmed in block:", receipt["blockNumber"])


if __name__ == "__main__":
    main()

File: node-js-deposit-redeem-example/deposit.js
require("dotenv").config();
const { ethers } = require("ethers");

/**
 * IMPORTANT:
 * We intentionally access `.abi` from the JSON files because these files are
 * Hardhat-style artifacts, not plain ABI arrays.
 */
const rwaAbi = require("./abis/RWAToken.json").abi;
const whitelistAbi = require("./abis/Whitelist.json").abi;
const feeAbi = require("./abis/RWAFee.json").abi;

/**
 * Load and validate environment variables used by this flow.
 */
function getEnv() {
  const env = {
    RPC_URL: process.env.RPC_URL,
    PRIVATE_KEY: process.env.PRIVATE_KEY,
    RWA_TOKEN_ADDRESS: process.env.RWA_TOKEN_ADDRESS,
    WHITELIST_ADDRESS: process.env.WHITELIST_ADDRESS,
    FEE_ADDRESS: process.env.FEE_ADDRESS || "",
    APPROVE_MAX: String(process.env.APPROVE_MAX || "false").toLowerCase() === "true",
    ENTRY_FEE_TIMING: Number(process.env.ENTRY_FEE_TIMING || 0),
  };

  if (!env.RPC_URL || !env.PRIVATE_KEY || !env.RWA_TOKEN_ADDRESS || !env.WHITELIST_ADDRESS) {
    throw new Error("Missing required env values. Please check .env.example");
  }

  return env;
}

/**
 * Resolve the fee contract address.
 *
 * Priority:
 * 1. use FEE_ADDRESS from env if provided
 * 2. fallback to vault.rwaFee() if env value is empty
 */
async function resolveFeeAddress(vault, env) {
  if (env.FEE_ADDRESS) {
    return env.FEE_ADDRESS;
  }

  const feeAddress = await vault.rwaFee();
  if (!feeAddress || feeAddress === ethers.ZeroAddress) {
    throw new Error("Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address.");
  }

  return feeAddress;
}

/**
 * Main deposit flow.
 *
 * User input = desired assets before entry fee.
 *
 * We calculate:
 * - entryFee = feeOnRaw(desiredAssets, ENTRY)
 * - totalSpend = desiredAssets + entryFee
 *
 * Then we:
 * - check whitelist
 * - check balance >= totalSpend
 * - check allowance >= totalSpend
 * - optionally approve
 * - call deposit(totalSpend, receiver)
 *
 * previewDeposit(totalSpend) is shown so integrators can confirm the share output.
 */
async function main() {
  const desiredAssetsInput = process.argv[2];
  if (!desiredAssetsInput) {
    throw new Error("Usage: node deposit.js <desiredAssetsBeforeFee>");
  }

  const env = getEnv();

  const provider = new ethers.JsonRpcProvider(env.RPC_URL);
  const wallet = new ethers.Wallet(env.PRIVATE_KEY, provider);

  const vault = new ethers.Contract(env.RWA_TOKEN_ADDRESS, rwaAbi, wallet);
  const whitelist = new ethers.Contract(env.WHITELIST_ADDRESS, whitelistAbi, wallet);

  const user = await wallet.getAddress();

  console.log("=== DEPOSIT FLOW (WITH ENTRY FEE) ===");
  console.log("User:", user);
  console.log("Vault:", env.RWA_TOKEN_ADDRESS);
  console.log("Whitelist:", env.WHITELIST_ADDRESS);

  // 1) Check whitelist
  const isWhitelisted = await whitelist.balanceOf(user, 1); // Token ID 1 is the relevant whitelist token
  console.log("Whitelisted:", isWhitelisted);
  if (!isWhitelisted) {
    throw new Error("User is not whitelisted.");
  }

  // 2) Load vault asset token
  const assetAddress = await vault.asset();
  const asset = new ethers.Contract(
    assetAddress,
    [
      "function decimals() view returns (uint8)",
      "function symbol() view returns (string)",
      "function balanceOf(address) view returns (uint256)",
      "function allowance(address,address) view returns (uint256)",
      "function approve(address,uint256) returns (bool)"
    ],
    wallet
  );

  const assetDecimals = await asset.decimals();
  const assetSymbol = await asset.symbol();
  const desiredAssets = ethers.parseUnits(desiredAssetsInput, assetDecimals);

  // 3) Resolve and load fee contract
  const feeAddress = await resolveFeeAddress(vault, env);
  const feeContract = new ethers.Contract(feeAddress, feeAbi, wallet);

  const entryFee = await feeContract.feeOnRaw(desiredAssets, env.ENTRY_FEE_TIMING);
  const totalSpend = desiredAssets + entryFee;
  const previewShares = await vault.previewDeposit(totalSpend);

  const shareDecimals = await vault.decimals();
  const shareSymbol = await vault.symbol();

  console.log("Fee contract:", feeAddress);
  console.log("Desired assets before fee:", desiredAssetsInput, assetSymbol);
  console.log("Entry fee:", ethers.formatUnits(entryFee, assetDecimals), assetSymbol);
  console.log("Total spend sent to deposit():", ethers.formatUnits(totalSpend, assetDecimals), assetSymbol);
  console.log("previewDeposit(totalSpend):", ethers.formatUnits(previewShares, shareDecimals), shareSymbol);

  // 4) Check balance and allowance using TOTAL SPEND (assets + fee)
  const assetBalance = await asset.balanceOf(user);
  const allowance = await asset.allowance(user, env.RWA_TOKEN_ADDRESS);

  console.log("Wallet asset balance:", ethers.formatUnits(assetBalance, assetDecimals), assetSymbol);
  console.log("Allowance to vault:", ethers.formatUnits(allowance, assetDecimals), assetSymbol);

  if (assetBalance < totalSpend) {
    throw new Error("Not enough asset balance for desiredAssets + entryFee.");
  }

  if (allowance < totalSpend) {
    const approveAmount = env.APPROVE_MAX ? ethers.MaxUint256 : totalSpend;
    console.log("Allowance is not enough, sending approve()...");
    const approveTx = await asset.approve(env.RWA_TOKEN_ADDRESS, approveAmount);
    console.log("Approve tx:", approveTx.hash);
    await approveTx.wait();
    console.log("Approve confirmed.");
  } else {
    console.log("Allowance is already enough.");
  }

  // 5) Deposit total spend (desired assets + entry fee)
  console.log("Sending deposit(totalSpend, receiver)...");
  const tx = await vault.deposit(totalSpend, user);
  console.log("Deposit tx:", tx.hash);
  const receipt = await tx.wait();
  console.log("Deposit confirmed in block:", receipt.blockNumber);
}

main().catch((error) => {
  console.error("Deposit flow failed:", error.shortMessage || error.message || error);
  process.exit(1);
});
